include/circt/Scheduling/Scheduler.h

```diff
@@ -0,0 +1,253 @@
+//===- Scheduler.h - Common interface for scheduling algorithms -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Static scheduling algorithms for use in HLS flows all solve similar problems.
+// The classes in this file serve as an interface between clients and algorithm
+// implementations, and model a basic scheduling problem and some commonly used
+// extensions (e.g. modulo scheduling). This includes problem-specific utility
+// and verification methods.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SCHEDULING_SCHEDULER_H
+#define CIRCT_SCHEDULING_SCHEDULER_H
+
+#include "circt/Support/LLVM.h"
+
+#include "mlir/Support/StorageUniquer.h"
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SetVector.h"
+
+namespace circt {
+namespace sched {
+
+class Dependence;
+class OperatorType;
+
+/// This class models the most basic scheduling problem.
+///
+/// A problem instance is comprised of:
+///
+///  - *Operations*: The vertices in the data-flow graph to be scheduled. We use
+///    the abbreviation `op` in the scheduling API.
+///  - *Dependences*: The edges in the data-flow graph to be scheduled, modeling
+///    a precedence relation between the involved operations. Individual operand
+///    and result indices can be distinguished. Abbreviation: `dep`
+///  - *Operator types*: An abstraction of the characteristics of the target
```

Comment:
What is the rationale for modeling operator types on top of the `Operation`s themselves? Why not just schedule the lowered operations themselves, lowered to the appropriate level first?

Reply:
If this is intended to describe the static `Op` instead of the dynamic `Operation`, look at `AbstractOperation`.

Reply:
If I am understanding the talk correctly, these operator types are the lower-level that will actually carry out the operations being scheduled, and can have properties of their own. If that's correct, I think I agree with @teqdruid that this kind of sounds like you could model them as `Operation`s. An `Operation` has a name and we use an `Operation *` to represent an opaque handle in lots of places.

I guess let me ask a question slightly differently: what kind of operator types could you forsee that _wouldn't_ make sense to model as operations in a dialect?

Reply:
My rationale for modeling the the operator types explicitly is:
- they group together characteristics (latency, blocking time, propagation delay, ...) that otherwise would need to be attached to every operation.
- they are an elegant way to express that a subset of operations compete for the same resource. For example, let's say the target datapath will have two add-operators, and we want to execute operations *a*, *b* and *c* on them. In this scenario, we'd model one operator type "adder" with a limit of 2, and link it to *a*, *b*, *c*.
- they enable a scheduler to choose from a set of operator types (e.g. the small & slow vs. the large & fast multiplier).

Using the `AbstractOperation` to represent the operator type might be too inflexible. A `DivFOp` could be linked to different operator types based on the float type (f32 -> latency = 32, f64 -> latency = 64), or you could have a combined AddSub operator that would be used by `AddIOp`s and `SubIOp`s.

Using `Operation *`s as handles for the operator types would work great for "heavy things" like an FP operation, where you'd have a `hw.module[.extern|.generated]` that represents the operator type, but not so well for operations that just map to HDL primitives.

My `OperatorType` class gives clients the most freedom for modeling their operator libraries, but uses an extra class and the storage uniquer machinery just for representing a simple handle.

As I type this, I wonder if I should just use a StringRef? That would give us the good bits of the three alternatives above, at the cost of more expensive hashing when accessing the operator type property maps.

Reply:
I think I'm starting to understand why it makes sense to model these outside the Operation/AbstractOperation world. Giving clients freedom in this case seems like a good thing. Thanks for the explanation.

> As I type this, I wonder if I should just use a StringRef?

What about an `Identifier`? Those are interned within a context, and you could use the `containingOp`s context to get `Identifiers`, and then the lookups should be more efficient.

Reply:
Let me rephrase your response to ensure that I'm understanding it: the reason for `OperatorType` is that there is a many-to-many mapping of operations to possible implementations of said operations wherein the scheduler may want to reason about resource contention?

If I'm interpreting you correctly, I _think_ it is reasonable to require clients model the possible implementations as operations themselves. E.g. an `AddSubOp` to model your resource which is capable of mapping both adds and subs into it. You could then use the `AddSubOp`s `AbstractOperation` instead of your `OperationType`. You could use it as a key for lookups (as you do with `OperationType` now), have default latencies (for well known operations e.g. in the comb dialect), and allow "implementation" operations to define their latencies as inherent properties.

I suspect this would be useful down the road as well -- for instance, a scheduler could output a mapping of `Operation*` to `AbstractOperation` showing which implementation it selected -- from there it would be trivial to lower it. Alternatively, the scheduler could actually instantiate the "implementation" `AbstractOperation`s into the `Operation*` and provide a mapping of `Operation*`s to schedule into the implementing `Operation*`s on which they should run and the schedule itself. A later 'multiplexing' pass could then wire up the correct multiplexing logic.

Does this sound reasonable?

Reply:
> Let me rephrase your response to ensure that I'm understanding it: the reason for `OperatorType` is that there is a many-to-many mapping of operations to possible implementations of said operations wherein the scheduler may want to reason about resource contention?

Correct, though I should note that in the beginning, we'll only support a many (Operations)-to-one (OperatorType) mapping. I do have an ILP-based scheduler to solve the many-to-many-problem, but it is only applicable if the number of ops to choose for is small. Developing a good and more scalable heuristic is what I'll look into next, research-wise.
 
> [...] Does this sound reasonable?

Makes a lot of sense.
- Operation *i* (to be scheduled) = `mlir::Operation`
- Operator type *q* = `mlir::AbstractOperation`
- Operator (an instance of *q*, executing *i* in a specific timestep) = `mlir::Operation`, i.e. instance of the `AbstractOperation`

This would require that clients model their operator library as `Operation`s, which might become quite tedious if they have modules with parameterized bit widths, for example. There was an ODM talk about [dynamic definition](https://llvm.discourse.group/t/open-mlir-meeting-6-17-2021-dynamic-definition-of-dialects-ops-types-attributes/3670) of dialects and ops (which I haven't watched yet), but maybe that'll become handy here.

The infra you outline in the last paragraph sounds amazingly useful.

Nevertheless, the scheduler class' interface currently is deliberately low-level with its imperative API (`addOperation`, `setLatency`, ...), and in this context, I think the `Identifier` representation of operator types would be good fit.

We can build the infrastructure you propose on top of that API (operator type's ID is the `AbstractOperation`'s name). Over time, we can remove the indirection via the `Identifier` (algorithms won't care, as its just a handle), and gradually design the necessary `[AbstractOperation|Op]Interface`s to query problem parameters directly from the target ops (instead of having someone `set(...)` them).

@teqdruid Would the `Identifier` approach acceptable to you to, as a short term solution to get the scheduling library off the ground?

Reply:
> This would require that clients model their operator library as `Operation`s, which might become quite tedious if they have modules with parameterized bit widths, for example.

Yeah, I could see that. Especially if they want to lower to a set of extern modules. What we really want is an AbstractOperation which can optionally store MLIR attributes to be added to the operation to be created _and_ could provide information about the implementation (e.g. latency).

> @teqdruid Would the Identifier approach acceptable to you to, as a short term solution to get the scheduling library off the ground?

Yeah, I'm OK with that and the flat API for now. We should plan on iterating it.

Reply:
> Yeah, I'm OK with that and the flat API for now. We should plan on iterating it.

Great. Yes, definitely, I'm curious to see how this evolves.